// @auth0/nextjs-auth0/dist/errors/my-account-errors.js
import { SdkError } from "./sdk-error.js";
/**
 * Error class representing a connect account request error.
 */
export class MyAccountApiError extends SdkError {
    constructor({ type, title, detail, status, validationErrors }) {
        super(`${title}: ${detail}`);
        this.name = "MyAccountApiError";
        this.code = "my_account_api_error";
        this.type = type;
        this.title = title;
        this.detail = detail;
        this.status = status;
        this.validationErrors = validationErrors;
    }
}
/**
 * Enum representing error codes related to the connect account flow.
 */
export var ConnectAccountErrorCodes;
(function (ConnectAccountErrorCodes) {
    /**
     * The session is missing.
     */
    ConnectAccountErrorCodes["MISSING_SESSION"] = "missing_session";
    /**
     * Failed to initiate the connect account flow.
     */
    ConnectAccountErrorCodes["FAILED_TO_INITIATE"] = "failed_to_initiate";
    /**
     * Failed to complete the connect account flow.
     */
    ConnectAccountErrorCodes["FAILED_TO_COMPLETE"] = "failed_to_complete";
})(ConnectAccountErrorCodes || (ConnectAccountErrorCodes = {}));
/**
 * Error class representing a connect account error.
 */
export class ConnectAccountError extends SdkError {
    constructor({ code, message, cause }) {
        super(message);
        this.name = "ConnectAccountError";
        this.code = code;
        this.cause = cause;
    }
}
